File: benchmark-suite/nodejs/benchmark.js
#!/usr/bin/env node

/**
 * DDEX Suite Node.js Performance Benchmark
 * Tests both parser and builder across multiple file sizes
 */

const fs = require('fs');
const path = require('path');
const os = require('os');
const { performance } = require('perf_hooks');
const { Worker, isMainThread, parentPort, workerData } = require('worker_threads');

// Import DDEX Suite components
let DDEXParser, DDEXBuilder;
try {
    // Try to import from installed packages first
    const parserPkg = require('ddex-parser');
    DDEXParser = parserPkg.DDEXParser || parserPkg.default || parserPkg;
    
    const builderPkg = require('ddex-builder');
    DDEXBuilder = builderPkg.DDEXBuilder || builderPkg.default || builderPkg;
    
    console.log('📦 Loaded from npm packages');
} catch (e) {
    console.log('📦 Using local bindings...');
    try {
        // Fallback to local packages
        const parserPkg = require('../../packages/ddex-parser/bindings/node');
        DDEXParser = parserPkg.DDEXParser || parserPkg.default || parserPkg;
        
        const builderPkg = require('../../packages/ddex-builder/bindings/node');
        DDEXBuilder = builderPkg.DDEXBuilder || builderPkg.default || builderPkg;
    } catch (e2) {
        console.error('❌ Could not load DDEX Suite bindings:', e2.message);
        process.exit(1);
    }
}

class NodejsBenchmarkSuite {
    constructor() {
        console.log('DDEXParser type:', typeof DDEXParser);
        console.log('DDEXBuilder type:', typeof DDEXBuilder);
        
        this.parser = new DDEXParser();
        this.builder = DDEXBuilder ? new DDEXBuilder() : null;
        this.results = [];
    }

    getMemoryUsage() {
        const mem = process.memoryUsage();
        return {
            rss: mem.rss,
            heapUsed: mem.heapUsed,
            heapTotal: mem.heapTotal,
            external: mem.external,
        };
    }

    async benchmarkFile(filePath) {
        const fileName = path.basename(filePath);
        const stats = fs.statSync(filePath);
        const fileSize = stats.size;
        
        console.log(`📊 Benchmarking ${fileName} (${(fileSize / 1024).toFixed(1)}KB)...`);

        const memoryBefore = this.getMemoryUsage();
        let result = {
            fileName,
            fileSize,
            parseTime: 0,
            buildTime: 0,
            memoryUsage: {
                before: memoryBefore,
                after: null,
                peak: memoryBefore
            },
            success: false,
            error: null
        };

        try {
            // Parse benchmark
            const parseStart = performance.now();
            const xmlContent = fs.readFileSync(filePath, 'utf-8');
            const parseResult = await this.parser.parse(xmlContent);
            const parseTime = performance.now() - parseStart;
            result.parseTime = parseTime;

            const memoryAfterParse = this.getMemoryUsage();
            result.memoryUsage.peak = {
                rss: Math.max(memoryBefore.rss, memoryAfterParse.rss),
                heapUsed: Math.max(memoryBefore.heapUsed, memoryAfterParse.heapUsed)
            };

            // Build benchmark (if available)
            let buildTime = 0;
            try {
                if (this.builder && parseResult && parseResult.flat) {
                    const buildStart = performance.now();
                    // Create a simple build request from parsed data
                    const buildRequest = {
                        releases: parseResult.flat.releases || [],
                        resources: parseResult.flat.resources || [],
                        deals: parseResult.flat.deals || []
                    };
                    const buildResult = await this.builder.build(buildRequest);
                    buildTime = performance.now() - buildStart;
                    result.buildTime = buildTime;
                }
            } catch (buildError) {
                // Build might fail due to API differences, but that's okay for performance testing
                console.log(`  ⚠️  Build skipped: ${buildError.message}`);
            }

            result.memoryUsage.after = this.getMemoryUsage();
            result.success = true;

            const memUsageMB = (result.memoryUsage.peak.heapUsed - memoryBefore.heapUsed) / (1024 * 1024);
            console.log(`  ✅ Parse: ${parseTime.toFixed(2)}ms, Build: ${buildTime.toFixed(2)}ms, Memory: ${memUsageMB.toFixed(1)}MB`);

        } catch (error) {
            result.error = error.message;
            result.memoryUsage.after = this.getMemoryUsage();
            console.log(`  ❌ Failed: ${error.message}`);
        }

        return result;
    }

    async runParallelBenchmark(filePath, numWorkers) {
        console.log(`🚀 Testing parallel processing with ${numWorkers} workers...`);
        
        const xmlContent = fs.readFileSync(filePath, 'utf-8');
        const workerPromises = [];
        
        const start = performance.now();
        
        for (let i = 0; i < numWorkers; i++) {
            const workerPromise = new Promise((resolve, reject) => {
                const worker = new Worker(__filename, {
                    workerData: { xmlContent, workerId: i }
                });
                
                const workerStart = performance.now();
                worker.on('message', (result) => {
                    if (result.success) {
                        resolve(performance.now() - workerStart);
                    } else {
                        resolve(0);
                    }
                });
                
                worker.on('error', () => resolve(0));
                worker.on('exit', (code) => {
                    if (code !== 0) resolve(0);
                });
            });
            
            workerPromises.push(workerPromise);
        }
        
        const results = await Promise.all(workerPromises);
        const totalTime = performance.now() - start;
        const avgTime = results.filter(t => t > 0).reduce((a, b) => a + b, 0) / results.length;
        
        console.log(`  Total time: ${totalTime.toFixed(2)}ms, Average per worker: ${avgTime.toFixed(2)}ms`);
        return results;
    }

    generateReport() {
        console.log('\n🎯 NODE.JS PERFORMANCE BENCHMARK REPORT');
        console.log('='.repeat(80));
        
        let successfulParses = 0;
        let totalParseTime = 0;
        let totalBuildTime = 0;
        let totalMemoryUsage = 0;

        console.log(
            'File'.padEnd(15) +
            'Size'.padStart(8) +
            'Parse(ms)'.padStart(12) +
            'Build(ms)'.padStart(12) +
            'Memory(MB)'.padStart(12) +
            'Status'.padStart(10)
        );
        console.log('-'.repeat(80));

        for (const result of this.results) {
            const sizeStr = result.fileSize > 1024 * 1024 
                ? `${(result.fileSize / (1024 * 1024)).toFixed(1)}MB`
                : `${Math.round(result.fileSize / 1024)}KB`;
            
            const memUsage = result.memoryUsage.after && result.memoryUsage.before
                ? (result.memoryUsage.after.heapUsed - result.memoryUsage.before.heapUsed) / (1024 * 1024)
                : 0;

            const status = result.success ? '✅' : '❌';
            
            console.log(
                result.fileName.padEnd(15) +
                sizeStr.padStart(8) +
                result.parseTime.toFixed(2).padStart(12) +
                result.buildTime.toFixed(2).padStart(12) +
                memUsage.toFixed(1).padStart(12) +
                status.padStart(10)
            );

            if (result.success) {
                successfulParses++;
                totalParseTime += result.parseTime;
                totalBuildTime += result.buildTime;
                totalMemoryUsage += memUsage;
            }
        }

        console.log('-'.repeat(80));
        console.log('📈 PERFORMANCE SUMMARY:');
        console.log(`  Successful parses: ${successfulParses}/${this.results.length}`);

        if (successfulParses > 0) {
            const avgParse = totalParseTime / successfulParses;
            const avgBuild = totalBuildTime / successfulParses;
            const avgMemory = totalMemoryUsage / successfulParses;
            
            console.log(`  Average parse time: ${avgParse.toFixed(2)}ms`);
            console.log(`  Average build time: ${avgBuild.toFixed(2)}ms`);
            console.log(`  Average memory usage: ${avgMemory.toFixed(1)}MB`);
        }

        // Check targets
        const largeFileResults = this.results.filter(r => r.fileSize >= 10 * 1024 * 1024 && r.success);
        if (largeFileResults.length > 0) {
            const avg10mbParse = largeFileResults.reduce((sum, r) => sum + r.parseTime, 0) / largeFileResults.length;
            console.log(`  Average 10MB+ parse time: ${avg10mbParse.toFixed(2)}ms (target: <50ms) ${avg10mbParse < 50 ? '✅' : '❌'}`);
        }

        // Performance throughput
        const totalData = this.results.filter(r => r.success).reduce((sum, r) => sum + r.fileSize, 0);
        const totalTime = totalParseTime / 1000; // Convert to seconds
        if (totalTime > 0) {
            const throughputMBs = (totalData / (1024 * 1024)) / totalTime;
            console.log(`  Overall throughput: ${throughputMBs.toFixed(1)} MB/s`);
        }
    }

    async run() {
        console.log('🚀 DDEX Suite Node.js Performance Benchmark');
        console.log('='.repeat(60));

        // Test files
        const testFiles = [
            '../test-data/1kb.xml',
            '../test-data/5kb.xml',
            '../test-data/10kb.xml',
            '../test-data/50kb.xml',
            '../test-data/100kb.xml',
            '../test-data/500kb.xml',
            '../test-data/1mb.xml',
            '../test-data/5mb.xml',
            '../test-data/10mb.xml',
            '../test-data/25mb.xml',
        ].map(f => path.join(__dirname, f));

        console.log('\n📊 Testing parse performance...');
        for (const filePath of testFiles) {
            if (fs.existsSync(filePath)) {
                const result = await this.benchmarkFile(filePath);
                this.results.push(result);
            }
        }

        // Parallel processing test
        const mediumFile = testFiles.find(f => fs.existsSync(f) && f.includes('1mb'));
        if (mediumFile) {
            console.log('\n🔄 Testing parallel processing capabilities...');
            await this.runParallelBenchmark(mediumFile, 4);
            await this.runParallelBenchmark(mediumFile, 8);
        }

        // Generate report
        this.generateReport();
        
        console.log('\n🎉 Node.js benchmark completed!');
    }
}

// Worker thread handler
if (!isMainThread) {
    const { xmlContent, workerId } = workerData;
    
    (async () => {
        try {
            const parser = new DDEXParser();
            const parseResult = await parser.parse(xmlContent);
            parentPort.postMessage({ success: true, workerId });
        } catch (error) {
            parentPort.postMessage({ success: false, error: error.message, workerId });
        }
    })();
} else {
    // Main thread - run benchmark
    const suite = new NodejsBenchmarkSuite();
    suite.run().catch(console.error);
}

File: benchmark-suite/python/benchmark.py
#!/usr/bin/env python3
"""
DDEX Suite Python Performance Benchmark
Tests parser performance across multiple file sizes
"""

import os
import sys
import time
import psutil
import threading
import concurrent.futures
from pathlib import Path

# Try to import DDEX Parser
try:
    import ddex_parser
    print("📦 Loaded ddex-parser from PyPI")
except ImportError:
    try:
        # Fallback to local development build
        sys.path.insert(0, str(Path(__file__).parent.parent.parent / "packages" / "ddex-parser" / "bindings" / "python"))
        import ddex_parser
        print("📦 Loaded ddex-parser from local bindings")
    except ImportError:
        print("❌ Could not load ddex-parser")
        sys.exit(1)

class PythonBenchmarkSuite:
    def __init__(self):
        self.parser = ddex_parser.DDEXParser()
        self.results = []
        self.process = psutil.Process()

    def get_memory_usage(self):
        """Get current memory usage in bytes"""
        memory_info = self.process.memory_info()
        return {
            'rss': memory_info.rss,
            'vms': memory_info.vms,
        }

    def benchmark_file(self, file_path):
        """Benchmark parsing a single file"""
        file_name = os.path.basename(file_path)
        file_size = os.path.getsize(file_path)
        
        print(f"📊 Benchmarking {file_name} ({file_size / 1024:.1f}KB)...")

        memory_before = self.get_memory_usage()
        result = {
            'file_name': file_name,
            'file_size': file_size,
            'parse_time': 0,
            'memory_usage': {
                'before': memory_before,
                'after': None,
                'peak': memory_before
            },
            'success': False,
            'error': None
        }

        try:
            # Parse benchmark
            with open(file_path, 'r', encoding='utf-8') as f:
                xml_content = f.read()
            
            start_time = time.perf_counter()
            parse_result = self.parser.parse(xml_content)
            end_time = time.perf_counter()
            parse_time = (end_time - start_time) * 1000  # Convert to milliseconds
            result['parse_time'] = parse_time

            memory_after = self.get_memory_usage()
            result['memory_usage']['after'] = memory_after
            result['memory_usage']['peak'] = {
                'rss': max(memory_before['rss'], memory_after['rss']),
                'vms': max(memory_before['vms'], memory_after['vms'])
            }
            
            result['success'] = True

            mem_usage_mb = (memory_after['rss'] - memory_before['rss']) / (1024 * 1024)
            print(f"  ✅ Parse: {parse_time:.2f}ms, Memory: {mem_usage_mb:.1f}MB")

        except Exception as error:
            result['error'] = str(error)
            result['memory_usage']['after'] = self.get_memory_usage()
            print(f"  ❌ Failed: {error}")

        return result

    def run_parallel_benchmark(self, file_path, num_workers):
        """Test parallel processing with multiple workers"""
        print(f"🚀 Testing parallel processing with {num_workers} workers...")
        
        with open(file_path, 'r', encoding='utf-8') as f:
            xml_content = f.read()
        
        def parse_worker(worker_id):
            try:
                parser = ddex_parser.DDEXParser()
                start_time = time.perf_counter()
                result = parser.parse(xml_content)
                end_time = time.perf_counter()
                return (end_time - start_time) * 1000
            except Exception:
                return 0

        start_time = time.perf_counter()
        
        with concurrent.futures.ThreadPoolExecutor(max_workers=num_workers) as executor:
            futures = [executor.submit(parse_worker, i) for i in range(num_workers)]
            results = [future.result() for future in concurrent.futures.as_completed(futures)]
        
        total_time = (time.perf_counter() - start_time) * 1000
        avg_time = sum(r for r in results if r > 0) / len([r for r in results if r > 0])
        
        print(f"  Total time: {total_time:.2f}ms, Average per worker: {avg_time:.2f}ms")
        return results

    def generate_report(self):
        """Generate performance report"""
        print('\n🎯 PYTHON PERFORMANCE BENCHMARK REPORT')
        print('=' * 80)
        
        successful_parses = 0
        total_parse_time = 0
        total_memory_usage = 0

        print(f"{'File':<15} {'Size':>8} {'Parse(ms)':>12} {'Memory(MB)':>12} {'Status':>10}")
        print('-' * 67)

        for result in self.results:
            if result['file_size'] > 1024 * 1024:
                size_str = f"{result['file_size'] / (1024 * 1024):.1f}MB"
            else:
                size_str = f"{result['file_size'] // 1024}KB"
            
            mem_usage = 0
            if result['memory_usage']['after'] and result['memory_usage']['before']:
                mem_usage = (result['memory_usage']['after']['rss'] - 
                           result['memory_usage']['before']['rss']) / (1024 * 1024)

            status = '✅' if result['success'] else '❌'
            
            print(f"{result['file_name']:<15} {size_str:>8} {result['parse_time']:>12.2f} "
                  f"{mem_usage:>12.1f} {status:>10}")

            if result['success']:
                successful_parses += 1
                total_parse_time += result['parse_time']
                total_memory_usage += mem_usage

        print('-' * 67)
        print('📈 PERFORMANCE SUMMARY:')
        print(f"  Successful parses: {successful_parses}/{len(self.results)}")

        if successful_parses > 0:
            avg_parse = total_parse_time / successful_parses
            avg_memory = total_memory_usage / successful_parses
            
            print(f"  Average parse time: {avg_parse:.2f}ms")
            print(f"  Average memory usage: {avg_memory:.1f}MB")

        # Check targets
        large_file_results = [r for r in self.results if r['file_size'] >= 10 * 1024 * 1024 and r['success']]
        if large_file_results:
            avg_10mb_parse = sum(r['parse_time'] for r in large_file_results) / len(large_file_results)
            status = '✅' if avg_10mb_parse < 50 else '❌'
            print(f"  Average 10MB+ parse time: {avg_10mb_parse:.2f}ms (target: <50ms) {status}")

        # Performance throughput
        total_data = sum(r['file_size'] for r in self.results if r['success'])
        total_time_sec = total_parse_time / 1000  # Convert to seconds
        if total_time_sec > 0:
            throughput_mbs = (total_data / (1024 * 1024)) / total_time_sec
            print(f"  Overall throughput: {throughput_mbs:.1f} MB/s")

    def run(self):
        """Run the benchmark suite"""
        print('🚀 DDEX Suite Python Performance Benchmark')
        print('=' * 60)

        # Test files
        test_files = [
            '../test-data/1kb.xml',
            '../test-data/5kb.xml',
            '../test-data/10kb.xml',
            '../test-data/50kb.xml',
            '../test-data/100kb.xml',
            '../test-data/500kb.xml',
            '../test-data/1mb.xml',
            '../test-data/5mb.xml',
            '../test-data/10mb.xml',
            '../test-data/25mb.xml',
        ]

        print('\n📊 Testing parse performance...')
        for file_name in test_files:
            file_path = Path(__file__).parent / file_name
            if file_path.exists():
                result = self.benchmark_file(file_path)
                self.results.append(result)

        # Parallel processing test
        medium_file = None
        for file_name in test_files:
            file_path = Path(__file__).parent / file_name
            if file_path.exists() and '1mb' in file_name:
                medium_file = file_path
                break

        if medium_file:
            print('\n🔄 Testing parallel processing capabilities...')
            self.run_parallel_benchmark(medium_file, 4)
            self.run_parallel_benchmark(medium_file, 8)

        # Generate report
        self.generate_report()
        
        print('\n🎉 Python benchmark completed!')

if __name__ == '__main__':
    suite = PythonBenchmarkSuite()
    suite.run()